Unrecognized self-referencing @scope/package

# Bug Report

<!--
  Please fill in each section completely. Thank you!
-->

### 🔎 Search Terms

- self-referencing
- self-referencing package
- node12
- module resolution

### 🕗 Version & Regression Information

TypeScript 4.5 Beta and TypeScript Nightly

### 🙁 Actual behavior

Given the next package.json:
```json
{
    "name": "@scope/package",
    "type": "module",
    "main": "./compiled/src/index.cjs",
    "exports": {
        "import": "./compiled/src/index.js",
        "require": "./compiled/src/index.cjs"
    },
    ...
}
```
And the following tsconfig.json:
```json
{
    "module": "node12",
    ...
}
```
The following code does not compile:
```ts
import * as src from "@scope/package"
```
The obtained error is `Cannot find module '@scope/package' or its corresponding type declarations.`

### 🙂 Expected behavior

TSC currently compile self-referencing package that has unscoped names. TSC should also compile self-referencing [scoped packages](https://docs.npmjs.com/cli/v6/using-npm/scope/) such as `@scope/package`.

--- a/src/compiler/moduleNameResolver.ts
+++ b/src/compiler/moduleNameResolver.ts
@@ -1724,11 +1724,15 @@ namespace ts {
         if (!scope || !scope.packageJsonContent.exports) {
             return undefined;
         }
-        const parts = getPathComponents(moduleName); // unrooted paths should have `""` as their 0th entry
-        if (scope.packageJsonContent.name !== parts[1]) {
+        if (typeof scope.packageJsonContent.name !== "string") {
             return undefined;
         }
-        const trailingParts = parts.slice(2);
+        const parts = getPathComponents(moduleName); // unrooted paths should have `""` as their 0th entry
+        const nameParts = getPathComponents(scope.packageJsonContent.name);
+        if (!every(nameParts, (p, i) => parts[i] === p)) {
+            return undefined;
+        }
+        const trailingParts = parts.slice(nameParts.length);
         return loadModuleFromExports(scope, extensions, !length(trailingParts) ? "." : `.${directorySeparator}${trailingParts.join(directorySeparator)}`, state, cache, redirectedReference);
     }
 

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).errors.txt
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).errors.txt
@@ -1,0 +1,24 @@
+tests/cases/conformance/node/index.cts(2,23): error TS1471: Module '@scope/package' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.
+
+
+==== tests/cases/conformance/node/index.ts (0 errors) ====
+    // esm format file
+    import * as self from "@scope/package";
+    self;
+==== tests/cases/conformance/node/index.mts (0 errors) ====
+    // esm format file
+    import * as self from "@scope/package";
+    self;
+==== tests/cases/conformance/node/index.cts (1 errors) ====
+    // cjs format file
+    import * as self from "@scope/package";
+                          ~~~~~~~~~~~~~~~~
+!!! error TS1471: Module '@scope/package' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.
+    self;
+==== tests/cases/conformance/node/package.json (0 errors) ====
+    {
+        "name": "@scope/package",
+        "private": true,
+        "type": "module",
+        "exports": "./index.js"
+    }

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).js
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).js
@@ -1,0 +1,63 @@
+//// [tests/cases/conformance/node/nodePackageSelfNameScoped.ts] ////
+
+//// [index.ts]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.mts]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.cts]
+// cjs format file
+import * as self from "@scope/package";
+self;
+//// [package.json]
+{
+    "name": "@scope/package",
+    "private": true,
+    "type": "module",
+    "exports": "./index.js"
+}
+
+//// [index.js]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.mjs]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.cjs]
+"use strict";
+var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
+    if (k2 === undefined) k2 = k;
+    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
+}) : (function(o, m, k, k2) {
+    if (k2 === undefined) k2 = k;
+    o[k2] = m[k];
+}));
+var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
+    Object.defineProperty(o, "default", { enumerable: true, value: v });
+}) : function(o, v) {
+    o["default"] = v;
+});
+var __importStar = (this && this.__importStar) || function (mod) {
+    if (mod && mod.__esModule) return mod;
+    var result = {};
+    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
+    __setModuleDefault(result, mod);
+    return result;
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+// cjs format file
+const self = __importStar(require("@scope/package"));
+self;
+
+
+//// [index.d.ts]
+export {};
+//// [index.d.mts]
+export {};
+//// [index.d.cts]
+export {};

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).symbols
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).symbols
@@ -1,0 +1,24 @@
+=== tests/cases/conformance/node/index.ts ===
+// esm format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.ts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.ts, 1, 6))
+
+=== tests/cases/conformance/node/index.mts ===
+// esm format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.mts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.mts, 1, 6))
+
+=== tests/cases/conformance/node/index.cts ===
+// cjs format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.cts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.cts, 1, 6))
+

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).types
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=node12).types
@@ -1,0 +1,24 @@
+=== tests/cases/conformance/node/index.ts ===
+// esm format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+
+=== tests/cases/conformance/node/index.mts ===
+// esm format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+
+=== tests/cases/conformance/node/index.cts ===
+// cjs format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).errors.txt
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).errors.txt
@@ -1,0 +1,24 @@
+tests/cases/conformance/node/index.cts(2,23): error TS1471: Module '@scope/package' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.
+
+
+==== tests/cases/conformance/node/index.ts (0 errors) ====
+    // esm format file
+    import * as self from "@scope/package";
+    self;
+==== tests/cases/conformance/node/index.mts (0 errors) ====
+    // esm format file
+    import * as self from "@scope/package";
+    self;
+==== tests/cases/conformance/node/index.cts (1 errors) ====
+    // cjs format file
+    import * as self from "@scope/package";
+                          ~~~~~~~~~~~~~~~~
+!!! error TS1471: Module '@scope/package' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.
+    self;
+==== tests/cases/conformance/node/package.json (0 errors) ====
+    {
+        "name": "@scope/package",
+        "private": true,
+        "type": "module",
+        "exports": "./index.js"
+    }

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).js
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).js
@@ -1,0 +1,63 @@
+//// [tests/cases/conformance/node/nodePackageSelfNameScoped.ts] ////
+
+//// [index.ts]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.mts]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.cts]
+// cjs format file
+import * as self from "@scope/package";
+self;
+//// [package.json]
+{
+    "name": "@scope/package",
+    "private": true,
+    "type": "module",
+    "exports": "./index.js"
+}
+
+//// [index.js]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.mjs]
+// esm format file
+import * as self from "@scope/package";
+self;
+//// [index.cjs]
+"use strict";
+var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
+    if (k2 === undefined) k2 = k;
+    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
+}) : (function(o, m, k, k2) {
+    if (k2 === undefined) k2 = k;
+    o[k2] = m[k];
+}));
+var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
+    Object.defineProperty(o, "default", { enumerable: true, value: v });
+}) : function(o, v) {
+    o["default"] = v;
+});
+var __importStar = (this && this.__importStar) || function (mod) {
+    if (mod && mod.__esModule) return mod;
+    var result = {};
+    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
+    __setModuleDefault(result, mod);
+    return result;
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+// cjs format file
+const self = __importStar(require("@scope/package"));
+self;
+
+
+//// [index.d.ts]
+export {};
+//// [index.d.mts]
+export {};
+//// [index.d.cts]
+export {};

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).symbols
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).symbols
@@ -1,0 +1,24 @@
+=== tests/cases/conformance/node/index.ts ===
+// esm format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.ts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.ts, 1, 6))
+
+=== tests/cases/conformance/node/index.mts ===
+// esm format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.mts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.mts, 1, 6))
+
+=== tests/cases/conformance/node/index.cts ===
+// cjs format file
+import * as self from "@scope/package";
+>self : Symbol(self, Decl(index.cts, 1, 6))
+
+self;
+>self : Symbol(self, Decl(index.cts, 1, 6))
+

--- a/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).types
+++ b/tests/baselines/reference/nodePackageSelfNameScoped(module=nodenext).types
@@ -1,0 +1,24 @@
+=== tests/cases/conformance/node/index.ts ===
+// esm format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+
+=== tests/cases/conformance/node/index.mts ===
+// esm format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+
+=== tests/cases/conformance/node/index.cts ===
+// cjs format file
+import * as self from "@scope/package";
+>self : typeof self
+
+self;
+>self : typeof self
+

--- a/tests/cases/conformance/node/nodePackageSelfNameScoped.ts
+++ b/tests/cases/conformance/node/nodePackageSelfNameScoped.ts
@@ -1,0 +1,21 @@
+// @module: node12,nodenext
+// @declaration: true
+// @filename: index.ts
+// esm format file
+import * as self from "@scope/package";
+self;
+// @filename: index.mts
+// esm format file
+import * as self from "@scope/package";
+self;
+// @filename: index.cts
+// cjs format file
+import * as self from "@scope/package";
+self;
+// @filename: package.json
+{
+    "name": "@scope/package",
+    "private": true,
+    "type": "module",
+    "exports": "./index.js"
+}

